doc/design/diagrams/request_response/typed_api.puml

```diff
@@ -0,0 +1,100 @@
+@startuml
```

Comment:
For Larry I wrote myself also some minimalistic client server thingy. Here I think we orient us to much on the existing publisher and subscriber port which are in this context a little bit cumbersome to use.

Here I drafted a code snippet on how I as a user would like to use the typed API.
```cpp
// Client side
// usage 1, wait directly for response of a request
Client<int, int> client;
auto response = client.sendCopyOf(123).waitForResponse();

// usage 2, acquire response from request handle
Client<int, int> client;
auto request = client.loan();
*request = 123;
request.send();
if ( request.hasResponse() ) // do stuff
// or wait here for a response
auto response request.waitForResponse();


// Server side.
// usage 1 respond via request (introduce answer method)
Server<int, int> server;
server.take().and_then([&](auto & request){
    request.answerWithCopyOf(*request + 5);
});

// usage 2 request comes with an already loan chunk for the response
Server<int, int> server;
auto request = server.take();
*request.getResponse().getDataPtr() = *request + 5;
*request.getResponse().send();
```

Maybe it is out of scope of this design document but for the public API (typed) I would start with a (pseude) code example to see how it look and feels. And I would optimize the typed API for easy usability - if the user would like to have very fine grained control over everything they can use the untyped API.

Reply:
I think there will be one or two iterations over the API once we write a few examples.

While `client.send().wait()` looks nice, I think that's what we created the `WaitSet`/`Listener` for. So these should be used for blocking calls.

The `request.hasResponse` is also not always feasible since it is possible to send multiple requests and get out of order responses. This has also to be taken into account.

The `answerWithCopy`, etc. convenience methods can of course be added. I omitted them because they are not crucial for a first implementation.

The `request.getResponse().send()` only works for non fire and forget request, but could be done by `getResponse` returning an optional. I also thought about this as alternative to `loan` and kind of like it. The response already needs to know the server in order for the sample deleter, so this might work.

@Indra5196 what do you think about the suggestions of elfenpiff?

Reply:
Hi @elBoberido @elfenpiff 
Yup. Sounds good for a start!!

Reply:
We might also restrict the the server options.

@Indra5196 do we need to support fire and forget in parallel to response? I mean, does a server need to do fire and forget and response at the same time?

Reply:
Like in the case where we are handling multiple clients, we might need to do this. Will this require locking of the request handler on the server side?